schema/Readme.md

```diff
@@ -359,207 +364,183 @@ Prisma core provides explicit support for all 3 relation types and connectors mu
+  customer  Customer?  @virtual
   name      String
 }
 
 model Customer {
   id       Int     @id
-  user     User?
+  user     User    @unique
   address  String
 }
 ```
 
-For 1:1 relationships, it doesn't matter which side you store the foreign key. Therefore Prisma has a convention that the foreign key is added to the model who
-appears first alphanumerically. In the example above, that's the `Customer` model.
-
-Under the hood, the models looks like this:
-
-| **users** |         |
-| --------- | ------- |
-| id        | integer |
-| name      | text    |
+Under the hood, the models looks like this in the database:
 
-| **customers** |         |
-| ------------- | ------- |
-| id            | integer |
-| user          | integer |
-| address       | text    |
+```sql
+create table User (
+  id integer primary key,
+  name text not null
+);
 
-You **may** omit either `User.customer` or `Customer.user` and the relationship will remain intact. This makes either the back-relation or the forward-relation
-optional. If one side of the relation is missing, Prisma implies the field name based on the name of the model it is pointing to.
+create table Customer (
+  id integer primary key,
+  user integer not null references User(id) unique,
+  address text not null
+);
+```
 
-If you're introspecting an existing database and the foreign key does not follow the alphanumeric convention, then we'll use the
-`@relation(_ name: String?, references: Identifier[]?, onDelete: OnDeleteEnum?)` attribute to clarify.
+Here's another valid example where the database backing has been reversed. Additionally, both sides are now optional:
 
 ```groovy
 model User {
   id        Int        @id
-  customer  Customer?  @relation(references: id)
+  customer  Customer?  @unique
   name      String
 }
 
 model Customer {
   id       Int     @id
-  user     User?
+  user     User?   @virtual
   address  String
 }
 ```
 
-##### One-to-Many (1:N) Relationships
-
-A writer can have multiple blogs.
+```sql
+create table User (
+  id integer primary key,
+  customer integer not null references User(id) unique,
+  name text not null
+);
 
-```groovy
-model Writer {
-  id      Int     @id
-  blogs   Blog[]
-}
-
-model Blog {
-  id      Int     @id
-  author  Writer
-}
+create table Customer (
+  id integer primary key,
+  address text not null
+);
 ```
 
-- `Blog.author`: points to the primary key on writer
-
-Connectors for relational databases will implement this as two tables with a foreign-key constraint on the blogs table:
-
-| **writers** |         |
-| ----------- | ------- |
-| id          | integer |
-
-| **blogs** |         |
-| --------- | ------- |
-| id        | integer |
-| author    | integer |
+###### Rules for One-to-One (1:1) Relationships
 
-###### Implied Has-Many
+1. You **must** include both sides of the relation. `User.customer` and `Customer.user`
+2. The virtual back-relation **must be** optional. `User.customer` must be `Customer?`
```

Comment:
What's the reasoning behind this? This basically means I can't have a 1:1 relation where both sides are required?

Reply:
@nikolasburk I think this is somewhat outdated at this point, though the use-case of having both required is limited imo. Here's why I wrote this:

**1. There's not really anyway to enforce both sides of a relation being required in SQL.** 

One-to-one relations typically look like this:

```sql
create table User (
  id integer primary key,
  customer integer not null references User(id) unique,
  name text not null
);

create table Customer (
  id integer primary key,
  address text not null
);
```

Well, I could have a Customer that doesn't have a User, right? There's nothing on the Customer side that explicitly links the Customer to a User, the data is store on the User's side. Now, we might be able to also store data on the Customer side in this case, but the goal of this PR is to make things explicit & not magical.

**2. Required on both sides is harder to use**

If you insert a Customer, you'll get an error because you don't have a User. If you insert a User, you'll get an error because you don't have a Customer. 

The only way to avoid this would be to do both inside a transaction. I'm not sure if the SQL databases do provide this as a guarantee but, we could definitely offer it at the Prisma level.

Reply:
> 2. Required on both sides is harder to use

Yup, I totally get this point and tbh I'm not sure how common this use case actually is... as you said, the only way to actually create records in a required 1:1 relation is by using a transaction, in Photon's case a nested write would be the appropriate way to do this. 

However, I'm not sure whether we should make these kinds of assumptions about the potential use cases. I could see this as a pretty nice feature for people that actually do have this requirement for whatsoever reason. In that case, we would even provide more value than if they were using plain SQL, since as you mentioned in your other point, it's not possible to do this with just SQL but you need to add this guarantee on the application level then. So, here would be a nice value add for people that have this use case to get this out-of-the box. Just my 2cents :)

Reply:
@matthewmueller I strongly disagree with this point:

> the use-case of having both required is very limited in my opinion

It's true that this is not possible to ensure on a DB level (for most DBs), this is a very valuable **guarantee we add on the Prisma level**.

We should always distinguish between functionality that's exposed (or sometimes even intentionally hidden) coming from the **DB level** vs a feature that's added on the **Prisma level**. Features added on the Prisma level are basically common enough scenarios that the user would typically implement in their application but Prisma offers as convenience/value-add out-of-the-box.

Tactically, I'd be fine with **delaying** this feature a bit but categorically think it's very important to implement common scenarios as Prisma features (_in this case, an added guarantee_).

Reply:
Just adding some facts:
* Yes we have a bunch of code to support required 1:1 relations as a Prisma level guarentee.
* I understand that this feature feels weird. But i can for sure say that people are make use of this. Just think about a Customer and an Address for example. With DB level only features it would be hard to ensure this kind of invariant.

Reply:
To clarify my understanding and make sure we're all on the same page. In order to create a relationship like this, you'd need to do something like:

```js
photon.customer.create({
  data: {
    name: "Alice",
    address: {
      street: "Wahoo Lane",
      zip: 15123
    }
  }
})
```

And that this is not possible:

```js
const customer = await photon.customer.create({
  name: "Alice"
})

const address = await photon.address.create({
    data: {
      street: "Wahoo Lane",
      zip: 15123  
    },
    link: {
      customer: customer.id
    }
})
```